runtime/parachains/src/disputes.rs

```diff
@@ -0,0 +1,752 @@
+					// Local disputes don't count towards spam.
+
+					weight += T::DbWeight::get().reads_writes(1, 1);
+					continue;
+				}
+
+				// mildly punish all validators involved. they've failed to make
+				// data available to others, so this is most likely spam.
+				SpamSlots::<T>::mutate(session_index, |spam_slots| {
+					let spam_slots = match spam_slots {
+						Some(ref mut s) => s,
+						None => return,
+					};
+
+					// also reduce spam slots for all validators involved, if the dispute was unconfirmed.
+					// this does open us up to more spam, but only for validators who are willing
+					// to be punished more.
+					//
+					// it would be unexpected for any change here to occur when the dispute has not concluded
+					// in time, as a dispute guaranteed to have at least one honest participant should
+					// conclude quickly.
+					let participating = decrement_spam(spam_slots, &dispute);
+
+					// Slight punishment as these validators have failed to make data available to
+					// others in a timely manner.
+					T::PunishValidators::punish_inconclusive(
+						session_index,
+						participating.iter_ones().map(|i| ValidatorIndex(i as _)),
+					);
+				});
+
+				weight += T::DbWeight::get().reads_writes(2, 2);
+			}
+		}
+
+		weight
+	}
+
+	/// Called by the initalizer to finalize the disputes module.
+	pub(crate) fn initializer_finalize() { }
+
+	/// Called by the initalizer to note a new session in the disputes module.
+	pub(crate) fn initializer_on_new_session(notification: &SessionChangeNotification<T::BlockNumber>) {
+		let config = <configuration::Pallet<T>>::config();
+
+		if notification.session_index <= config.dispute_period + 1 {
+			return
+		}
+
+		let pruning_target = notification.session_index - config.dispute_period - 1;
+
+		LastPrunedSession::<T>::mutate(|last_pruned| {
+			if let Some(last_pruned) = last_pruned {
+				for to_prune in *last_pruned + 1 ..= pruning_target {
+					<Disputes<T>>::remove_prefix(to_prune);
+					<Included<T>>::remove_prefix(to_prune);
+					SpamSlots::<T>::remove(to_prune);
+				}
+			}
+
+			*last_pruned = Some(pruning_target);
+		});
+	}
+
+	/// Handle sets of dispute statements corresponding to 0 or more candidates.
+	pub(crate) fn provide_multi_dispute_data(statement_sets: MultiDisputeStatementSet)
+		-> Result<Vec<(SessionIndex, CandidateHash)>, DispatchError>
+	{
+		let config = <configuration::Pallet<T>>::config();
+
+		// Deduplicate.
+		{
+			let mut targets: Vec<_> = statement_sets.iter()
+				.map(|set| (set.candidate_hash.0, set.session))
+				.collect();
+
+			targets.sort();
+
+			let submitted = targets.len();
+			targets.dedup();
+
+			ensure!(submitted == targets.len(), Error::<T>::DuplicateDisputeStatementSets);
+		}
+
+		let mut fresh = Vec::with_capacity(statement_sets.len());
+		for statement_set in statement_sets {
+			let dispute_target = (statement_set.session, statement_set.candidate_hash);
+			if Self::provide_dispute_data(&config, statement_set)? {
```

Comment:
in case this fail, a dispatch error is returned, but the previous successful call modified the storage.

I don't know exactly in which context provide_multi_dispute_data is used, but this contradict the usual rule of `check first then writes`, and in case of failure, the fresh written statement are not returned by the function.

So I guess it is a mistake

Reply:
Because the `enter` inherent is required for block validity in Polkadot (see `on_finalize` implementation), it is OK for us to bypass the typical requirement of not changing storage before failing. That allows us to make the parachains implementation more efficient.